extensions/mssql/src/credentialstore/nativeCredentialService.ts

```diff
@@ -0,0 +1,145 @@
+			private _passwordsMigrated: boolean;
+
+			constructor(client: SqlOpsDataClient) {
+				super(context, client);
+				this._secretStorage = context.extensionContext.secrets;
+				this._passwordsMigrated = context.extensionContext.globalState.get(Utils.configPasswordsMigrated);
+				this._useNativeCredentialService = Utils.useNativeCredentialsEnabled();
+			}
+
+			override fillClientCapabilities(capabilities: ClientCapabilities): void {
+				Utils.ensure(Utils.ensure(capabilities, 'credentials')!, 'credentials')!.dynamicRegistration = true;
+			}
+
+			override async initialize(capabilities: ServerCapabilities): Promise<void> {
+				this.register(this.messages, {
+					id: UUID.generateUuid(),
+					registerOptions: undefined
+				});
+				if (this._useNativeCredentialService && !this._passwordsMigrated) {
+					const success = await this.migratePasswords();
+					this._passwordsMigrated = success;
+					await context.extensionContext.globalState.update(Utils.configPasswordsMigrated, success);
+				}
+			}
+
+			/**
+			 * Removes a credential for a given connection profile from sqltoolsservice
+			 * and adds it to secret storage
+			 */
+			private async cleanCredential(conn: azdata.connection.ConnectionProfile): Promise<boolean> {
+				const credentialId = Utils.formatCredentialId(conn);
+				// read password from every saved password connection
+				const credential = await this._client.sendRequest(Contracts.ReadCredentialRequest.type, { credentialId, password: undefined });
+				if (credential.password) {
+					const password = credential.password;
+					// save it in secret store
+					await this._secretStorage.store(credentialId, password);
+					// check if it's saved
+					const savedPassword = await this._secretStorage.get(credentialId);
+					if (savedPassword === password) {
+						// delete from tools service
+						const result = await this._client.sendRequest(Contracts.DeleteCredentialRequest.type,
+							{ credentialId, password: undefined });
+						return result;
+					} else {
+						return false;
+					}
+				} else {
+					return false;
+				}
+			}
+
+			/**
+			 *
+			 * @returns Migrates all saved credentials to the native credential system
+			 */
+			private async migratePasswords(): Promise<boolean> {
+				const connections = await azdata.connection.getConnections(false);
+				const savedPasswordConnections = connections.filter(conn => conn.savePassword === true);
+				for (let i = 0; i < savedPasswordConnections.length; i++) {
+					let conn = savedPasswordConnections[i];
+					await this.cleanCredential(conn);
+				}
+				await Utils.removeCredentialFile();
+				return true;
+			}
+
+			protected override registerProvider(options: any): Disposable {
+				let readCredential = async (credentialId: string): Promise<azdata.Credential> => {
+					if (this._useNativeCredentialService && this._passwordsMigrated) {
+						const password = await this._secretStorage.get(credentialId);
```

Comment:
It's unfortunate that this isn't going to apply for anyone using STS (such as .NET Interactive). What was the issue with doing it natively in C# code for all the platforms?